cmd/status/printers/table_printer.go

```diff
@@ -0,0 +1,385 @@
+)
+
+const (
+	// updateInterval defines how often the printer will update the UI.
+	updateInterval = 1 * time.Second
+)
+
+// printContentFunc defines the function type used by the printer to ask a ColumnDef
+// to output the appropriate content for a given resource and column.
+type printContentFunc func(w io.Writer, width int, resource *event.ResourceStatus) (int, error)
+
+// ColumnDef defines the properties of every supported column by the printer.
+type ColumnDef struct {
+	// name of the column. This is just a string that is unique among all the defined columns
+	name string
+	// header defines the title of the column that will be printed in the table.
+	header string
+	// width defines the width of the column
+	width int
+	// printContent defines a function that will be called by the printer
+	// to output information into the given cell in the table for a provided
+	// ResourceStatus. This function is responsible for trimming any content that
+	// is too wide and takes care of setting the color of anything that is printed.
+	printContent printContentFunc
+}
+
+var (
+	// columns contains a list of ColumnDefs, one for each of the columns that
+	// is supported by the printer.
+	columns = []ColumnDef{
+		{
+			// Column containing the namespace of the resource.
+			name:   "namespace",
+			header: "NAMESPACE",
+			width:  10,
+			printContent: func(w io.Writer, availableWidth int, resource *event.ResourceStatus) (int, error) {
+				namespace := resource.Identifier.Namespace
+				if len(namespace) > availableWidth {
+					namespace = namespace[:availableWidth]
+				}
+				_, err := fmt.Fprint(w, namespace)
+				return len(namespace), err
+			},
+		},
+		{
+			// Column containing the resource type and name. Currently it does not print group or version
+			// since those are rarely needed to uniquely distinguish two resources from each other. Just name
+			// and kind should be enough in almost all cases and saves space in the output.
+			name:   "resource",
+			header: "RESOURCE",
+			width:  40,
+			printContent: func(w io.Writer, availableWidth int, resource *event.ResourceStatus) (int, error) {
+				text := fmt.Sprintf("%s/%s", resource.Identifier.GroupKind.Kind, resource.Identifier.Name)
+				if len(text) > availableWidth {
+					text = text[:availableWidth]
+				}
+				_, err := fmt.Fprint(w, text)
+				return len(text), err
+			},
+		},
+		{
+			// Column containing the status of the resource as computed by the status and polling
+			// libraries.
+			name:   "status",
+			header: "STATUS",
+			width:  10,
+			printContent: func(w io.Writer, availableWidth int, resource *event.ResourceStatus) (int, error) {
+				s := resource.Status.String()
+				if len(s) > availableWidth {
+					s = s[:availableWidth]
+				}
+				color, setColor := colorForTableStatus(resource.Status)
+				var outputStatus string
+				if setColor {
+					outputStatus = sPrintWithColor(color, s)
+				} else {
+					outputStatus = s
+				}
```

Comment:
```
stat := s
if !setColor {
  stat = sPrintWithColor(color, s)
}
```

Reply:
So I tried something like this earlier, but I felt it ended up being worse. The challenge is that the code needs to keep track of the column width in the output. The ANSI characters are a challenge here since they do count when checking the length of a string in the program, but does not actually matter when we print. But I am interested in finding a cleaner solution here, but I would like to get back to it in a separate PR.